connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3468,6 +3554,26 @@ class Dataset[T] private[sql] (
     }
   }
 
+  // Visible for testing
+  private[sql] var cachedRemoteRelationID: Option[String] = None
+
+  override def finalize(): Unit = {
```

Comment:
Finalize has been deprecated in java 9 for good reasons. Let's use a `java.lang.ref.Cleaner` instead.

However high level I have a different concern. This assumes that the original dataframe is always in scope, however if that is being garbage collected any derived dataframe breaks. I think we should pin it to the original CachedRemoteRelation instead, that will be re-used in derived dataframes. The upside of this is that you don't have to introduce a new field.

Reply:
I also think this is an issue for python.

Reply:
For Python actually there's no alternative, and `__del__` isn't actually deprecated in Python. I read why it's deprecated in JDK (e.g., no guarantee on the order of `fianlize` invoke, etc.), and it won't affect this specific usecase but yeah let me switch it to `java.lang.ref.Cleaner`.

> This assumes that the original dataframe is always in scope, however if that is being garbage collected any derived dataframe breaks. I think we should pin it to the original CachedRemoteRelation instead, that will be re-used in derived dataframes.

Yeah, this is a good point. Let me fix it up.

Reply:
For Python, https://github.com/apache/spark/pull/46694